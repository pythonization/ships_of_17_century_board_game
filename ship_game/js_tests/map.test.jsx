import {
    expect, test, vi, describe,
    afterEach
} from 'vitest'
import renderer from 'react-test-renderer';
import {
    cleanup, render,
    fireEvent
} from '@testing-library/react';

import ShipMap, { MapCell } from '../static/ship_game/components/map/map.jsx'


const mock_callback = vi.fn()

describe('Render tests', () => {
    function render_cell(cell_d) {
        return renderer
            .create(<MapCell cell_d={cell_d} toggle_cell={mock_callback} />)
            .toJSON()
    }

    test('Cell (all 3 types)', () => {
        expect(
            render_cell(' ')
            // better use "asFragment()" instead of 'react-test-renderer'
        ).toMatchSnapshot();
        expect(
            render_cell('W')
            // better use "asFragment()" instead of 'react-test-renderer'
        ).toMatchSnapshot();
        expect(
            render_cell('P')
            // better use "asFragment()" instead of 'react-test-renderer'
        ).toMatchSnapshot();
    })

    test('Map (Readonly)', () => {
        expect(
            renderer
                .create(
                    <ShipMap
                        table={[
                            ['W', 'W', 'W'],
                            ['W', 'P', ' '],
                            ['W', ' ', ' '],
                        ]}
                        image='/not_existing_image.png'
                    />
                )
                .toJSON()
                // better use "asFragment()" instead of 'react-test-renderer'
        ).toMatchSnapshot();
    })

    test('Map (Editable)', () => {
        expect(
            renderer
                .create(
                    <ShipMap
                        table={[
                            ['W', 'W', 'W', 'W'],
                            ['W', 'W', 'W', 'W'],
                            ['W', 'W', 'P', ' '],
                            ['W', 'W', ' ', ' '],
                        ]}
                        image='/not_existing_image.png' on_edit_cell={mock_callback}
                    />
                )
                .toJSON()
                // better use "asFragment()" instead of 'react-test-renderer'
        ).toMatchSnapshot();
    })
})

describe('Tests with events handlers', () => {

    afterEach(() => {
        cleanup();
    });

    test('Cell', () => {

        // need this because <td> should be inside table structure
        const { container } = render(<table>
            <tbody>
                <tr>
                    <MapCell cell_d='W' toggle_cell={mock_callback} />
                </tr>
            </tbody>
        </table>)

        // Should not use "container" in test that simulate user's actions. Better not to use it at all.
        // but here it is easier to use it
        const tr_el = container.querySelector('td')
        fireEvent.mouseDown(tr_el)
        fireEvent.mouseEnter(tr_el)

        expect(
            mock_callback
        ).toHaveBeenCalledTimes(2);
    })
})