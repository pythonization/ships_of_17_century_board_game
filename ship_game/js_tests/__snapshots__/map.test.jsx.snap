// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Render tests > Cell (all 3 types) 1`] = `
<td
  onMouseDown={[MockFunction spy]}
  onMouseEnter={[MockFunction spy]}
  style={
    {
      "border": false,
    }
  }
/>
`;

exports[`Render tests > Cell (all 3 types) 2`] = `
<td
  onMouseDown={[MockFunction spy]}
  onMouseEnter={[MockFunction spy]}
  style={
    {
      "border": "1px solid black",
    }
  }
/>
`;

exports[`Render tests > Cell (all 3 types) 3`] = `
<td
  onMouseDown={[MockFunction spy]}
  onMouseEnter={[MockFunction spy]}
  style={
    {
      "border": "1px solid black",
    }
  }
>
  <i
    className="fa-solid fa-anchor"
  />
</td>
`;

exports[`Render tests > Map (Editable) 1`] = `
<table
  className="ship_map"
  onMouseDown={[Function]}
  style={
    {
      "backgroundImage": "url(\\"/not_existing_image.png\\")",
      "borderCollapse": "separate",
      "fontSize": "12.75vw",
      "width": "100vw",
    }
  }
>
  <tbody>
    <tr
      style={
        {
          "height": "calc(25vw - 2px)",
        }
      }
    >
      <td
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        style={
          {
            "border": "1px solid black",
          }
        }
      />
      <td
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        style={
          {
            "border": "1px solid black",
          }
        }
      />
      <td
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        style={
          {
            "border": "1px solid black",
          }
        }
      />
      <td
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        style={
          {
            "border": "1px solid black",
          }
        }
      />
    </tr>
    <tr
      style={
        {
          "height": "calc(25vw - 2px)",
        }
      }
    >
      <td
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        style={
          {
            "border": "1px solid black",
          }
        }
      />
      <td
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        style={
          {
            "border": "1px solid black",
          }
        }
      />
      <td
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        style={
          {
            "border": "1px solid black",
          }
        }
      />
      <td
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        style={
          {
            "border": "1px solid black",
          }
        }
      />
    </tr>
    <tr
      style={
        {
          "height": "calc(25vw - 2px)",
        }
      }
    >
      <td
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        style={
          {
            "border": "1px solid black",
          }
        }
      />
      <td
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        style={
          {
            "border": "1px solid black",
          }
        }
      />
      <td
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        style={
          {
            "border": "1px solid black",
          }
        }
      >
        <i
          className="fa-solid fa-anchor"
        />
      </td>
      <td
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        style={
          {
            "border": false,
          }
        }
      />
    </tr>
    <tr
      style={
        {
          "height": "calc(25vw - 2px)",
        }
      }
    >
      <td
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        style={
          {
            "border": "1px solid black",
          }
        }
      />
      <td
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        style={
          {
            "border": "1px solid black",
          }
        }
      />
      <td
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        style={
          {
            "border": false,
          }
        }
      />
      <td
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        style={
          {
            "border": false,
          }
        }
      />
    </tr>
  </tbody>
</table>
`;

exports[`Render tests > Map (Readonly) 1`] = `
<table
  className="ship_map"
  onMouseDown={[Function]}
  style={
    {
      "backgroundImage": "url(\\"/not_existing_image.png\\")",
      "borderCollapse": undefined,
      "fontSize": "17vw",
      "width": "100vw",
    }
  }
>
  <tbody>
    <tr
      style={
        {
          "height": "33.333333333333336vw",
        }
      }
    >
      <td
        style={
          {
            "border": "1px solid black",
          }
        }
      />
      <td
        style={
          {
            "border": "1px solid black",
          }
        }
      />
      <td
        style={
          {
            "border": "1px solid black",
          }
        }
      />
    </tr>
    <tr
      style={
        {
          "height": "33.333333333333336vw",
        }
      }
    >
      <td
        style={
          {
            "border": "1px solid black",
          }
        }
      />
      <td
        style={
          {
            "border": "1px solid black",
          }
        }
      >
        <i
          className="fa-solid fa-anchor"
        />
      </td>
      <td
        style={
          {
            "border": false,
          }
        }
      />
    </tr>
    <tr
      style={
        {
          "height": "33.333333333333336vw",
        }
      }
    >
      <td
        style={
          {
            "border": "1px solid black",
          }
        }
      />
      <td
        style={
          {
            "border": false,
          }
        }
      />
      <td
        style={
          {
            "border": false,
          }
        }
      />
    </tr>
  </tbody>
</table>
`;
